File: 백준/Silver/9012. 괄호/괄호.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())

for _ in range(n):
    s = deque()
    PS = input().rstrip()
    answer = 'YES'
    for x in PS:
        if x == '(':
            s.append(x)
        elif x == ')':
            if len(s) == 0:
                answer = 'NO'
                break
            else:
                s.pop()
    if len(s) != 0:
        answer = 'NO'
    print(answer)

File: 백준/Silver/9012. 괄호/괄호.js
const fs = require('fs');
const [n, ...input] = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

for (let i=0; i<n; i++) {
    let stack = [];
    let answer = 'YES';
    for (p of input[i]) {
        if (p == '(') {
            stack.push(p);
        } else {
            if (stack.length) {
                stack.pop();
            } else {
                answer = 'NO';
            }
        }
    }
    if (stack.length) {
        answer = 'NO';
    }
    console.log(answer);
}